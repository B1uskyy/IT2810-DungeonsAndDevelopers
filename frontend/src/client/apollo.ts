import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
});

const authLink = setContext((_, { headers }) => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  };
});

const client = new ApolloClient({
  link: authLink.concat(httpLink),
  devtools: {
    enabled: true,
  },
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          user: {
            keyArgs: ['id'],
            merge(existing = {}, incoming) {
              return { ...existing, ...incoming };
            },
          },
        },
      },
      User: {
        keyFields: ['id'],
        fields: {
          favoritedMonsters: {
            merge: false,
          },
          dungeonName: {
            merge: false,
          },
          class: {
            merge(_, incoming) {
              return incoming;
            },
          },

          race: {
            merge(_, incoming) {
              return incoming;
            },
          },
        },
      },
      Class: {
        keyFields: ['name'],
      },
      Race: {
        keyFields: ['name'],
      },
      Monster: {
        keyFields: ['id'],
      },
    },
  }),
});

export default client;
