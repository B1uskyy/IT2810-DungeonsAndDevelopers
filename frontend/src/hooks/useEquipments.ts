import { useEffect, useState } from 'react';

interface EquipmentItem {

  name: string;

}

function useEquipments() {
  const [data, setData] = useState<{
    results: EquipmentItem[];

  }>({
    results: [],

  });


  useEffect(() => {
    fetch('https://www.dnd5eapi.co/api/equipment/')
      .then((response) => response.json())
      .then((json) => setData(json))
      .catch((error) => console.log(error));
  }, []);

  return data.results;

}

export default useEquipments;
