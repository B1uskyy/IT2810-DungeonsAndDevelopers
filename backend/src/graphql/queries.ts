import { gql } from '@apollo/client';

export const GET_MONSTERS = gql`
    query GetMonsters($searchTerm: String, $offset: Int, $limit: Int, $types: [String!]) {
        monsters(searchTerm: $searchTerm, offset: $offset, limit: $limit, types: $types) {
            monsters {
                id
                name
                size
                type
                alignment
                hit_points
                image
            }
            totalMonsters
        }
    }
`;

export const GET_USER_DUNGEON = gql`
  query GetUserDungeon($userId: ID!) {
    user(id: $userId) {
      dungeonName
      favoritedMonsters {
        id
      }
    }
  }
`;

export const UPDATE_DUNGEON_NAME = gql`
  mutation UpdateDungeonName($userId: ID!, $dungeonName: String!) {
    updateDungeonName(userId: $userId, dungeonName: $dungeonName) {
      dungeonName
    }
  }
`;

export const GET_USER_FAVORITES = gql`
  query GetUserFavorites($userId: ID!) {
    user(id: $userId) {
      favoritedMonsters {
        id
        name
        size
        type
        alignment
        hit_points
        image
      }
    }
  }
`;

export const ADD_FAVORITE_MONSTER = gql`
  mutation AddFavoriteMonster($userId: ID!, $monsterId: ID!) {
    addFavoriteMonster(userId: $userId, monsterId: $monsterId) {
      favoritedMonsters {
        id
      }
    }
  }
`;

export const REMOVE_FAVORITE_MONSTER = gql`
  mutation RemoveFavoriteMonster($userId: ID!, $monsterId: ID!) {
    removeFavoriteMonster(userId: $userId, monsterId: $monsterId) {
      favoritedMonsters {
        id
      }
    }
  }
`;

export const GET_MONSTER_REVIEWS = gql`
  query GetMonsterReviews($monsterId: ID!) {
    monster(id: $monsterId) {
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;

export const ADD_REVIEW = gql`
  mutation AddReview($monsterId: ID!, $review: ReviewInput!) {
    addReview(monsterId: $monsterId, review: $review) {
      reviews {
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;

export const DELETE_REVIEW = gql`
  mutation DeleteReview($monsterId: ID!, $reviewId: ID!) {
    deleteReview(monsterId: $monsterId, reviewId: $reviewId) {
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;

export const UPDATE_REVIEW = gql`
  mutation UpdateReview($monsterId: ID!, $reviewId: ID!, $review: ReviewInput!) {
    updateReview(monsterId: $monsterId, reviewId: $reviewId, review: $review) {
      id
      user {
        id
        userName
      }
      difficulty
      description
      createdAt
    }
  }
`;

export const CHECK_USERNAME = gql`
    query checkUsername($userName: String!) {
        checkUsername(userName: $userName)
    }
`;

export const GET_USER_EQUIPMENT = gql`
  query GetUserEquipment($userId: ID!) {
    user(id: $userId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const ADD_EQUIPMENT_TO_CHARACTER = gql`
  mutation AddEquipmentToCharacter($userId: ID!, $equipmentId: ID!) {
    addEquipmentToCharacter(userId: $userId, equipmentId: $equipmentId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const REMOVE_EQUIPMENT_FROM_CHARACTER = gql`
  mutation RemoveEquipmentFromCharacter($userId: ID!, $equipmentId: ID!) {
    removeEquipmentFromCharacter(userId: $userId, equipmentId: $equipmentId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const CREATE_USER = gql`
  mutation createUser($userName: String!) {
    createUser(userName: $userName) {
      token
      user {
        id
        userName
        class {
          name
        }
        race {
          name
        }
        abilityScores {
          name
          score
        }
        equipments {
          name
        }
      }
    }
  }
`;

export const LOGIN_USER = gql`
  mutation loginUser($userName: String!) {
    loginUser(userName: $userName) {
      token
      user {
        id
        userName
        class {
          name
        }
        race {
          name
        }
        abilityScores {
          name
          score
        }
        equipments {
          name
        }
      }
    }
  }
`;
export const UPDATE_USER_CLASS = gql`
    mutation UpdateUserClass($userId: ID!, $classId:ID!){
        updateUserClass(userId:$userId, classId: $classId){
            id
            class{
                name
                index
                hit_die
            }
        }
    }
`;

export const GET_CLASSES = gql`
    query GetClasses($offset: Int, $limit: Int) {
        classes(offset: $offset, limit: $limit) {
            classes {
                id
                index
                name
                hit_die
                skills
            }
            totalClasses
        }
    }
`;

export const GET_USER_CLASS = gql`
    query GetUserClass($userId: ID!) {
        user(id: $userId) {
            class {
                id
                name
                index
                hit_die
            }
        }
    }
`;

export const UPDATE_USER_RACE = gql`
    mutation UpdateUserRace($userId: ID!, $raceId: ID!) {
        updateUserRace(userId: $userId, raceId: $raceId) {
            id
            race {
                name
                speed
                alignment
                size
                img
            }
        }
    }
`;

export const GET_RACES = gql`
    query GetRaces($offset: Int, $limit: Int) {
        races(offset: $offset, limit: $limit) {
            races {
                id
                index
                name
                speed
                alignment
                size
                img
            }
            totalRaces
        }
    }
`;

export const GET_USER_RACE = gql`
    query GetUserRace($userId: ID!) {
        user(id: $userId) {
            id
            race {
                id
                name
                speed
                alignment
                size
                img
            }
        }
    }
`;