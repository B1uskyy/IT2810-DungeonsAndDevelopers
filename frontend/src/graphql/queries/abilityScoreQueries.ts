import { gql } from '@apollo/client';

export const GET_ABILITYSCORES = gql`
  query GetAbilityScores($offset: Int, $limit: Int) {
    abilities(offset: $offset, limit: $limit) {
      abilities {
        index
        name
        skills
      }
      totalAbilities
    }
  }
`;
