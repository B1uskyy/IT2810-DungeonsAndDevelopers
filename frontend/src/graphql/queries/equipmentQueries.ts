import { gql } from '@apollo/client';

export const GET_EQUIPMENT_SUGGESTIONS = gql`
  query GetEquipmentSuggestions($searchTerm: String!) {
    equipments(searchTerm: $searchTerm, suggestionsOnly: true) {
      equipments {
        id
        name
      }
      totalCount
    }
  }
`;

export const GET_EQUIPMENTS = gql`
  query GetEquipments($searchTerm: String, $offset: Int, $limit: Int) {
    equipments(searchTerm: $searchTerm, offset: $offset, limit: $limit) {
      equipments {
        id
        index
        name
        category
        value
      }
      totalCount
    }
  }
`;
