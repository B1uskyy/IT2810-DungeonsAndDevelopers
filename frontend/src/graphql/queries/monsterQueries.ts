import { gql } from '@apollo/client';

export const GET_MONSTERS = gql`
  query GetMonsters(
    $searchTerm: String
    $offset: Int
    $limit: Int
    $types: [String!]
    $minHp: Int
    $maxHp: Int
    $sortOption: String
  ) {
    monsters(
      searchTerm: $searchTerm
      offset: $offset
      limit: $limit
      types: $types
      minHp: $minHp
      maxHp: $maxHp
      sortOption: $sortOption
    ) {
      monsters {
        id
        name
        size
        type
        alignment
        hit_points
        image
        averageDifficulty
        reviewsCount
      }
      totalMonsters
      minHp
      maxHp
    }
  }
`;

export const GET_MONSTER_TYPE_COUNTS = gql`
  query GetMonsterTypeCounts($minHp: Int, $maxHp: Int) {
    monsterTypeCounts(minHp: $minHp, maxHp: $maxHp) {
      type
      count
    }
  }
`;

export const GET_MONSTER_HP_RANGE = gql`
  query GetMonsterHpRange {
    monsterHpRange {
      minHp
      maxHp
    }
  }
`;

export const GET_MONSTER_SUGGESTIONS = gql`
  query GetMonsterSuggestions($searchTerm: String, $types: [String!], $minHp: Int, $maxHp: Int) {
    monsters(searchTerm: $searchTerm, types: $types, minHp: $minHp, maxHp: $maxHp) {
      monsters {
        id
        name
      }
    }
  }
`;

export const GET_MONSTER_REVIEWS = gql`
  query GetMonsterReviews($monsterId: ID!) {
    monster(id: $monsterId) {
      id
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;
