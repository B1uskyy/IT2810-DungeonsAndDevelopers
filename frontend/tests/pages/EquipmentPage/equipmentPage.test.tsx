import { render, screen, waitFor, within } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { MockedProvider } from '@apollo/client/testing';
import EquipmentPage from '../../../src/pages/subPages/equipmentPage.tsx';
import { GET_USER_EQUIPMENT, REMOVE_ALL_EQUIPMENTS } from '../../../../backend/src/graphql/queries.ts';
import { AuthContext } from '../../../src/context/AuthContext.tsx';
import { BrowserRouter } from 'react-router-dom';
import { GET_EQUIPMENTS } from '../../../src/hooks/useEquipments.ts';
import { expect } from 'vitest';

const mockShowToast = vi.fn();
vi.mock('../../../src/hooks/useToast.ts', () => ({
  useToast: () => ({
    showToast: mockShowToast,
  }),
}));

const mockEquipments = [
  { id: '1', index: '1', name: 'Helmet', category: 'Headgear', value: 4 },
  { id: '2', index: '2', name: 'Sword', category: 'Weapon', value: 5 },
];
const mockUserEquipments = [{ id: '1', index: '1', name: 'Helmet', category: 'Headgear', value: 4 }];
const mockUserId = '1';
const mockToken = 'mock-token'; // Mock token
const mockUserName = 'Mock User'; // Mock user name

const mocks = [
  {
    request: {
      query: GET_EQUIPMENTS,
      variables: { offset: 0, limit: 20 },
    },
    result: {
      data: {
        equipments: {
          equipments: mockEquipments,
          totalCount: 2,
        },
      },
    },
  },
  {
    request: {
      query: GET_USER_EQUIPMENT,
      variables: { userId: mockUserId },
    },
    result: {
      data: {
        user: {
          id: mockUserId,
          equipments: mockUserEquipments,
        },
      },
    },
  },
  {
    request: {
      query: REMOVE_ALL_EQUIPMENTS,
      variables: { userId: '1' },
    },
    result: {
      data: {
        removeAllEquipments: true,
      },
    },
  },
  {
    request: {
      query: GET_USER_EQUIPMENT,
      variables: { userId: mockUserId },
    },
    result: {
      data: {
        user: {
          id: mockUserId,
          equipments: [],
        },
      },
    },
  },
];

describe('EquipmentPage', () => {
  beforeEach(() => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AuthContext.Provider
          value={{ userId: mockUserId, token: mockToken, userName: mockUserName, login: vi.fn(), logout: vi.fn() }}
        >
          <BrowserRouter>
            <EquipmentPage />
          </BrowserRouter>
        </AuthContext.Provider>
      </MockedProvider>
    );
  });

  it('renders correctly and matches snapshot', async () => {
    await waitFor(() => {
      expect(screen.getByText('Equipments')).toBeInTheDocument();
    });
    expect(screen.queryByText('No equipments available')).toBeNull();

    expect(screen.getByRole('main')).toMatchSnapshot();
  });

  it('handles adding an equipment to user equipments', async () => {
    const user = userEvent.setup();
    const equipmentCard = await screen.findByText('Helmet');

    const checkBox = await within(equipmentCard).findByRole('checkbox');

    await user.click(checkBox);

    await waitFor(() =>
      expect(mockShowToast).toHaveBeenCalledWith({
        message: 'Helmet was added to your equipments',
        type: 'success',
        duration: 3000,
      })
    );
  });

  it('handles removing an equipment from user equipments', async () => {
    const user = userEvent.setup();
    const equipmentCard = screen.getByTestId('Helmet-equipment-card');
    const removeButton = within(equipmentCard).getByText('Remove from equipments');
    await user.click(removeButton);

    await waitFor(() =>
      expect(mockShowToast).toHaveBeenCalledWith({
        message: 'Helmet removed from equipments',
        type: 'info',
        duration: 5000,
      })
    );
  });

  it('removes all equipments', async () => {
    const user = userEvent.setup();
    const removeAllButton = screen.getByText('Remove All Equipments');
    await user.click(removeAllButton);

    await waitFor(() =>
      expect(mockShowToast).toHaveBeenCalledWith({
        message: 'All equipments removed',
        type: 'info',
        duration: 5000,
      })
    );
  });

  it('matches snapshot for paginated equipments', async () => {
    await waitFor(() => {
      expect(screen.getByTestId('pagination')).toBeInTheDocument();
    });
    expect(screen.getByRole('navigation')).toMatchSnapshot();
  });
});
