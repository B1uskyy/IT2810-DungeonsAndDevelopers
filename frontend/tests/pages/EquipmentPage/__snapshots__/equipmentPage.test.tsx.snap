// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EquipmentPage > renders correctly and matches snapshot 1`] = `
<main
  class="main before:bg-equipments"
>
  <div
    class="black-overlay"
  />
  <div
    class="wrapper py-20 min-w-[70%] flex gap-y-32 2xl:gap-0 mt-10 items-center justify-center"
  >
    <h1
      class="header"
    >
      Equipments
    </h1>
    <button
      class="text px-1 rounded-md bg-customRed hover:bg-transparent border-2 border-customRed hover:border-customRed hover:text-customRed transition-colors duration-200"
    >
      Remove All Equipments
    </button>
    <section
      class="w-full h-9/10"
    >
      <div
        class="grid xs:grid-cols-1 sm:grid-cols-2 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 3xl:grid-cols-5 gap-10 p-10 w-full h-full auto-rows-fr"
        style="opacity: 1; will-change: transform; transform: none;"
      >
        <section
          class="flex flex-col items-center justify-between gap-4 bg-black bg-opacity-80 py-10 px-4 rounded-xl w-full h-full min-w-[18vw] min-h-[30vh] text-center"
        >
          <h3
            class="text"
          >
            Helmet
          </h3>
          <p
            class="text"
          >
            Value: 4
          </p>
          <label
            class="relative block cursor-pointer transform transition-transform"
            style="transform: scale(1.5);"
          >
            <input
              checked=""
              type="checkbox"
            />
            <span
              class="absolute top-0 left-0 h-6 w-6 rounded bg-gray-300 transition-all duration-300 bg-red-600"
            />
            <span
              class="absolute left-[9px] top-[5px] w-[5px] h-[10px] border-white border-2 border-t-0 border-l-0 transform rotate-45"
            />
          </label>
        </section>
        <section
          class="flex flex-col items-center justify-between gap-4 bg-black bg-opacity-80 py-10 px-4 rounded-xl w-full h-full min-w-[18vw] min-h-[30vh] text-center"
        >
          <h3
            class="text"
          >
            Sword
          </h3>
          <p
            class="text"
          >
            Value: 5
          </p>
          <label
            class="relative block cursor-pointer transform transition-transform"
            style="transform: scale(1.5);"
          >
            <input
              type="checkbox"
            />
            <span
              class="absolute top-0 left-0 h-6 w-6 rounded bg-gray-300 transition-all duration-300 "
            />
          </label>
        </section>
      </div>
    </section>
    <section
      class="w-full flex justify-center items-center mt-5"
    >
      <div
        class="flex justify-between items-center w-[70vw] xl:w-[30vw] w-text-2xl xl:text-lg 2xl:text-md"
      >
        <button
          class="flex items-center hover:text-gray-400"
          disabled=""
        >
          <svg
            class="mr-2"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
            />
          </svg>
          Prev Page
        </button>
        <span
          class="w-2"
        >
          1
        </span>
        <button
          class="flex items-center hover:text-gray-400"
          disabled=""
        >
          Next Page
          <svg
            class="ml-2"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
            />
          </svg>
        </button>
      </div>
    </section>
  </div>
</main>
`;
